pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,149 +18,506 @@ import (
 	"context"
 	"fmt"
 	"math"
-	"strings"
 
-	"github.com/Pallinder/go-randomdata"
-	"github.com/mitchellh/hashstructure/v2"
+	"github.com/aws/karpenter/pkg/utils/apiobject"
+
+	"go.uber.org/multierr"
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/selection"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
-	"github.com/aws/karpenter/pkg/utils/functional"
 	"github.com/aws/karpenter/pkg/utils/pod"
+	"github.com/aws/karpenter/pkg/utils/sets"
 )
 
 type Topology struct {
 	kubeClient client.Client
+	topologies map[uint64]*TopologyGroup
+	// Anti-affinity works both ways (if a zone has a pod foo with anti-affinity to a pod bar, we can't schedule bar to
+	// that zone, even though bar has no anti affinity terms on it. For this to work, we need to separately track the
+	// topologies of pods with anti-affinity terms, so we can prevent scheduling the pods they have anti-affinity to
+	// in some cases.
+	inverseTopologies map[uint64]*TopologyGroup
+	constraints       *v1alpha5.Constraints
+}
+
+func NewTopology(kubeClient client.Client, constraints *v1alpha5.Constraints) *Topology {
+	return &Topology{
+		kubeClient:        kubeClient,
+		constraints:       constraints,
+		topologies:        map[uint64]*TopologyGroup{},
+		inverseTopologies: map[uint64]*TopologyGroup{},
+	}
 }
 
-// Inject injects topology rules into pods using supported NodeSelectors
-func (t *Topology) Inject(ctx context.Context, constraints *v1alpha5.Constraints, pods []*v1.Pod) error {
-	// Group pods by equivalent topology spread constraints
-	topologyGroups := t.getTopologyGroups(pods)
-	// Compute spread
-	for _, topologyGroup := range topologyGroups {
-		if err := t.computeCurrentTopology(ctx, constraints, topologyGroup); err != nil {
-			return fmt.Errorf("computing topology, %w", err)
+type matchingTopology struct {
+	topology              *TopologyGroup
+	controlsPodScheduling bool
+	selectsPod            bool
+}
+
+// Initialize scans the pods provided and creates topology groups for any topologies that we need to track based off of
+// topology spreads, affinities, and anti-affinities specified in the pods.
+func (t *Topology) Initialize(ctx context.Context, pods ...*v1.Pod) error {
+	var errs error
+	errs = multierr.Append(errs, t.trackExistingAntiAffinities(ctx))
+	for _, p := range pods {
+		// need to first ensure that we know all of the topology constraints.  This may require looking up
+		// existing pods in the running cluster to determine zonal topology skew.
+		if err := t.trackTopologySpread(ctx, p); err != nil {
```

Comment:
I typically try to avoid verb sprawl. In this case `track` `count` `update` `compute` `calculate` `warm` are all viable terms, and I'm sure there are many more I'm not thinking of. I try to minimize the nuance in verbs and instead push all of the context into the noun. This helps constrain the naming space, and results in more similar code from different developers, especially over time. It also helps guide encapsulation best practices, since if your interface has two things it needs to Add, it's a good hint to separate the abstraction.

It's obviously not perfect, but across this codebase we have a relatively small number of simple verbs `Create` `Delete` `Add` `Update` and then a set of "terms of art" from the domain like `Pack` and `Schedule`. I could imagine taking it a step further and replacing those terms of art e.g. `scheduler.Add`, `packing.Add`.

```suggestion
		if err := t.updateTopologySpread(ctx, p); err != nil {
```